### ISSUE TITLE
源码中未点击的广告变量j 相关的问题

### ISSUE THREAD

**Issue by wyljpn** [zhougr1993/DeepInterestNetwork#19]
你好。
我拜读了[Deep Interest Network for Click-Through Rate Prediction] 这篇文论文，并且下载、运行了你的这份源码。有些问题想请你赐教。

在build_dataset.py中，生成test_set（(reviewerID, hist, (最后点击的广告的id, 随机出的一个未点击的广告的id))），单条数据的格式是类似于这样：

> (0, [13179,17993,28326,29247],(62275,5940))

在input.py中给j赋值

> j.append(t[2][1])

所以我就理解成变量j是未点击的广告。
在model.py中的__init__方法，训练时的前向传播有使用到在指定历史下 未点击的广告id:

>  jc = tf.gather(cate_list, self.j)
        j_emb = tf.concat([
            tf.nn.embedding_lookup(item_emb_w, self.j),
            tf.nn.embedding_lookup(cate_emb_w, jc),
        ], axis=1)
        j_b = tf.gather(item_b, self.j)
        din_j = tf.concat([u_emb, j_emb], axis=-1)
        din_j = tf.layers.batch_normalization(inputs=din_j, name='b1', reuse=True)
        d_layer_1_j = tf.layers.dense(din_j, 80, activation=tf.nn.sigmoid, name='f1', reuse=True)
        # d_layer_1_j = dice(d_layer_1_j, name='dice_1_j')
        d_layer_2_j = tf.layers.dense(d_layer_1_j, 40, activation=tf.nn.sigmoid, name='f2', reuse=True)
        # d_layer_2_j = dice(d_layer_2_j, name='dice_2_j')
        d_layer_3_j = tf.layers.dense(d_layer_2_j, 1, activation=None, name='f3', reuse=True)
        d_layer_3_i = tf.reshape(d_layer_3_i, [-1])
        d_layer_3_j = tf.reshape(d_layer_3_j, [-1])
        x = i_b - j_b + d_layer_3_i - d_layer_3_j  # [B]
        self.logits = i_b + d_layer_3_i

那么问题来了，
1：在训练时使用未点击的广告的信息的目的是什么？是为了得到用户不感兴趣的信息？
2：变量x的作用是什么？
希望能得到回复，非常感谢！

**Comment by zhougr1993** [comment]
A1：未点击的item是我们通过随机采样生成的，在训练和测试过程中都作为负样本。如果没有负样本，那么我们设定的loss function没有意义，它会有一个很简单的解，就是全预估为正就好了。
A2：i_b和j_b分别是对应item_i和item_j的bias项，i_b+d_layer_3_i 和 j_b+d_layer_3_j分别是item_i和item_j最后一层sigmoid激活前的logits。sigmoid后的结果为点击率，由于sigmoid是单增函数，所以x = (i_b+d_layer_3_i) - (j_b + d_layer_3_j) 表达item_i的CTR预估值和item_j的CTR预估值的大小关系。

**Comment by wyljpn** [comment]
非常感谢您热心地回答我的问题。
重新看了一下论文和代码，理解了一些之前不懂的细节。
